File: packages/alpaca-option/src/chain.ts
import { expiration as timeExpiration } from '@alpaca/time';

import { canonicalContract, type ContractInput } from './contract';

type SnapshotLike = {
  contract?: ContractInput;
} | null | undefined;

type SnapshotCollection<T extends SnapshotLike> = {
  snapshots?: T[] | null;
} | null | undefined;

type NumericInput = number | string | null | undefined;

function coerceNumber(value: NumericInput): number | null {
  if (value == null) {
    return null;
  }

  const parsed = typeof value === 'number' ? value : Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

function strikeTolerance(value: NumericInput): number {
  const parsed = coerceNumber(value);
  return parsed != null && parsed >= 0 ? parsed : 0.01;
}

function matchesContract(
  contract: NonNullable<ReturnType<typeof canonicalContract>>,
  input: {
    occSymbol?: string;
    expirationDate?: string;
    strike?: NumericInput;
    optionRight?: string | null;
    strikeTolerance?: NumericInput;
  },
): boolean {
  if (input.occSymbol) {
    return contract.occ_symbol === input.occSymbol.trim().toUpperCase();
  }

  if (input.expirationDate && contract.expiration_date !== input.expirationDate) {
    return false;
  }

  if (input.optionRight && contract.option_right !== input.optionRight) {
    return false;
  }

  const strike = coerceNumber(input.strike);
  if (strike != null && Math.abs(contract.strike - strike) > strikeTolerance(input.strikeTolerance)) {
    return false;
  }

  return true;
}

export function listSnapshots<T extends SnapshotLike>(input: {
  chain?: SnapshotCollection<T>;
  occSymbol?: string;
  expirationDate?: string;
  strike?: NumericInput;
  optionRight?: string | null;
  strikeTolerance?: NumericInput;
}): T[] {
  const snapshots = input.chain?.snapshots ?? [];
  const results: T[] = [];

  for (const snapshot of snapshots) {
    const contract = canonicalContract(snapshot?.contract);
    if (!contract) {
      continue;
    }

    if (matchesContract(contract, input)) {
      results.push(snapshot);
    }
  }

  return results;
}

export function findSnapshot<T extends SnapshotLike>(input: {
  chain?: SnapshotCollection<T>;
  occSymbol?: string;
  expirationDate?: string;
  strike?: NumericInput;
  optionRight?: string | null;
  strikeTolerance?: NumericInput;
}): T | null {
  return listSnapshots(input)[0] ?? null;
}

export function expirationDates<T extends SnapshotLike>(input: {
  chain?: SnapshotCollection<T>;
  optionRight?: string | null;
  minCalendarDays?: NumericInput;
  maxCalendarDays?: NumericInput;
  now?: string;
}): Array<{ expirationDate: string; calendarDays: number }> {
  const minCalendarDays = coerceNumber(input.minCalendarDays);
  const maxCalendarDays = coerceNumber(input.maxCalendarDays);
  const seen = new Set<string>();
  const results: Array<{ expirationDate: string; calendarDays: number }> = [];

  for (const snapshot of input.chain?.snapshots ?? []) {
    const contract = canonicalContract(snapshot?.contract);
    if (!contract) {
      continue;
    }

    if (input.optionRight && contract.option_right !== input.optionRight) {
      continue;
    }

    if (seen.has(contract.expiration_date)) {
      continue;
    }

    const calendarDays = timeExpiration.calendarDays(contract.expiration_date, input.now);
    if (minCalendarDays != null && calendarDays < minCalendarDays) {
      continue;
    }
    if (maxCalendarDays != null && calendarDays > maxCalendarDays) {
      continue;
    }

    seen.add(contract.expiration_date);
    results.push({
      expirationDate: contract.expiration_date,
      calendarDays,
    });
  }

  return results.sort((left, right) => {
    if (left.calendarDays !== right.calendarDays) {
      return left.calendarDays - right.calendarDays;
    }
    return left.expirationDate.localeCompare(right.expirationDate);
  });
}

File: crates/alpaca-option/src/chain.rs
use serde::{Deserialize, Serialize};

use crate::error::{OptionError, OptionResult};
use crate::types::{OptionChain, OptionRight, OptionSnapshot};
use alpaca_time::expiration;

#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct SnapshotFilter<'a> {
    pub occ_symbol: Option<&'a str>,
    pub expiration_date: Option<&'a str>,
    pub strike: Option<f64>,
    pub option_right: Option<&'a str>,
    pub strike_tolerance: Option<f64>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ExpirationDate {
    pub expiration_date: String,
    pub calendar_days: i64,
}

fn strike_tolerance(value: Option<f64>) -> f64 {
    match value {
        Some(parsed) if parsed >= 0.0 => parsed,
        _ => 0.01,
    }
}

fn matches_contract(snapshot: &OptionSnapshot, filter: SnapshotFilter<'_>) -> bool {
    if let Some(occ_symbol) = filter.occ_symbol {
        return snapshot.contract.occ_symbol.eq_ignore_ascii_case(occ_symbol.trim());
    }

    if let Some(expiration_date) = filter.expiration_date {
        if snapshot.contract.expiration_date != expiration_date {
            return false;
        }
    }

    if let Some(option_right) = filter.option_right {
        let Ok(option_right) = OptionRight::from_str(option_right) else {
            return false;
        };
        if snapshot.contract.option_right != option_right {
            return false;
        }
    }

    if let Some(strike) = filter.strike {
        if (snapshot.contract.strike - strike).abs() > strike_tolerance(filter.strike_tolerance) {
            return false;
        }
    }

    true
}

pub fn list_snapshots<'a>(chain: &'a OptionChain, filter: SnapshotFilter<'_>) -> Vec<&'a OptionSnapshot> {
    chain
        .snapshots
        .iter()
        .filter(|snapshot| matches_contract(snapshot, filter))
        .collect()
}

pub fn find_snapshot<'a>(chain: &'a OptionChain, filter: SnapshotFilter<'_>) -> Option<&'a OptionSnapshot> {
    chain
        .snapshots
        .iter()
        .find(|snapshot| matches_contract(snapshot, filter))
}

pub fn expiration_dates(
    chain: &OptionChain,
    option_right: Option<&str>,
    min_calendar_days: Option<i64>,
    max_calendar_days: Option<i64>,
    now: Option<&str>,
) -> OptionResult<Vec<ExpirationDate>> {
    let option_right = option_right.map(OptionRight::from_str).transpose()?;
    let mut results = Vec::new();

    for snapshot in &chain.snapshots {
        if let Some(expected) = &option_right {
            if &snapshot.contract.option_right != expected {
                continue;
            }
        }

        if results
            .iter()
            .any(|item: &ExpirationDate| item.expiration_date == snapshot.contract.expiration_date)
        {
            continue;
        }

        let calendar_days = expiration::calendar_days(&snapshot.contract.expiration_date, now)
            .map_err(|error| OptionError::new(error.code, error.message))?;
        if let Some(min_calendar_days) = min_calendar_days {
            if calendar_days < min_calendar_days {
                continue;
            }
        }
        if let Some(max_calendar_days) = max_calendar_days {
            if calendar_days > max_calendar_days {
                continue;
            }
        }

        results.push(ExpirationDate {
            expiration_date: snapshot.contract.expiration_date.clone(),
            calendar_days,
        });
    }

    results.sort_by(|left, right| {
        left.calendar_days
            .cmp(&right.calendar_days)
            .then_with(|| left.expiration_date.cmp(&right.expiration_date))
    });

    Ok(results)
}
